File: src/services/taskManager.js
import { generateInstanceName, createPreemptibleVM, getVMStatus, deleteVM } from './vmManager.js';
import { uploadBuffer, fileExists, downloadFile } from './storage.js';
import { config } from '../config.js';
import { v4 as uuidv4 } from 'uuid';

/**
 * 任務狀態
 */
export const TaskStatus = {
  PENDING: 'PENDING',
  VM_CREATING: 'VM_CREATING',
  VM_RUNNING: 'VM_RUNNING',
  ANALYZING: 'ANALYZING',
  COMPLETED: 'COMPLETED',
  FAILED: 'FAILED',
  INTERRUPTED: 'INTERRUPTED',
};

// 記憶體中的任務狀態（生產環境建議使用 Redis 或資料庫）
const tasks = new Map();

/**
 * 建立新的分析任務
 * @param {string} userId - LINE 用戶 ID
 * @param {Buffer} sgfBuffer - SGF 檔案內容
 * @param {string} fileName - 檔案名稱
 * @returns {Promise<Object>} 任務資訊
 */
export async function createTask(userId, sgfBuffer, fileName) {
  const taskId = uuidv4();
  const timestamp = Date.now();
  
  // 上傳 SGF 檔案至 GCS
  const sgfPath = `sgf/${taskId}/${fileName}`;
  const sgfGcsPath = await uploadBuffer(sgfBuffer, sgfPath);

  // 建立任務記錄
  const task = {
    taskId,
    userId,
    fileName,
    sgfPath: sgfGcsPath,
    status: TaskStatus.PENDING,
    createdAt: timestamp,
    updatedAt: timestamp,
    instanceName: null,
    resultPath: null,
    error: null,
  };

  tasks.set(taskId, task);

  // 非同步啟動 VM 並執行分析
  startAnalysis(taskId).catch(error => {
    console.error(`Task ${taskId} failed:`, error);
    updateTaskStatus(taskId, TaskStatus.FAILED, { error: error.message });
  });

  return task;
}

/**
 * 啟動分析流程
 * @param {string} taskId - 任務 ID
 * @returns {Promise<void>}
 */
async function startAnalysis(taskId) {
  const task = tasks.get(taskId);
  if (!task) {
    throw new Error(`Task ${taskId} not found`);
  }

  try {
    // 產生 VM 實例名稱
    const instanceName = generateInstanceName();
    updateTaskStatus(taskId, TaskStatus.VM_CREATING, { instanceName });

    // 建立 Preemptible VM
    await createPreemptibleVM(instanceName, task.sgfPath);
    updateTaskStatus(taskId, TaskStatus.VM_RUNNING);

    // 開始監控任務狀態
    monitorTask(taskId, instanceName);
  } catch (error) {
    console.error(`Failed to start analysis for task ${taskId}:`, error);
    updateTaskStatus(taskId, TaskStatus.FAILED, { error: error.message });
    throw error;
  }
}

/**
 * 監控任務狀態
 * @param {string} taskId - 任務 ID
 * @param {string} instanceName - VM 實例名稱
 */
async function monitorTask(taskId, instanceName) {
  const maxWaitTime = 10 * 60 * 1000; // 10 分鐘
  const checkInterval = 5000; // 每 5 秒檢查一次
  const startTime = Date.now();

  const checkStatus = async () => {
    const task = tasks.get(taskId);
    if (!task) return;

    // 檢查是否超時
    if (Date.now() - startTime > maxWaitTime) {
      updateTaskStatus(taskId, TaskStatus.FAILED, {
        error: 'Task timeout',
      });
      await deleteVM(instanceName);
      return;
    }

    // 檢查結果檔案是否存在
    const resultPath = `results/${instanceName}/result.txt`;
    const statusPath = `results/${instanceName}/status.txt`;

    try {
      const resultExists = await fileExists(resultPath);
      const statusExists = await fileExists(statusPath);

      if (resultExists && statusExists) {
        // 任務完成
        updateTaskStatus(taskId, TaskStatus.COMPLETED, {
          resultPath: `gs://${config.storage.bucketName}/${resultPath}`,
        });
        
        // 清理 VM（如果還在運行）
        setTimeout(() => deleteVM(instanceName), 30000); // 30 秒後刪除
        return;
      }

      // 檢查 VM 狀態
      const vmStatus = await getVMStatus(instanceName);
      
      if (!vmStatus.exists || vmStatus.status === 'TERMINATED') {
        // VM 被中斷或已關閉
        if (resultExists) {
          // 有結果，可能是正常完成
          updateTaskStatus(taskId, TaskStatus.COMPLETED, {
            resultPath: `gs://${config.storage.bucketName}/${resultPath}`,
          });
        } else {
          // 沒有結果，可能是被中斷
          updateTaskStatus(taskId, TaskStatus.INTERRUPTED, {
            error: 'VM was preempted or terminated',
          });
          
          // 重試（最多重試 3 次）
          const retryCount = (task.retryCount || 0) + 1;
          if (retryCount < 3) {
            console.log(`Retrying task ${taskId}, attempt ${retryCount}`);
            task.retryCount = retryCount;
            await startAnalysis(taskId);
            return;
          }
        }
        return;
      }

      if (vmStatus.status === 'RUNNING') {
        updateTaskStatus(taskId, TaskStatus.ANALYZING);
      }

      // 繼續監控
      setTimeout(checkStatus, checkInterval);
    } catch (error) {
      console.error(`Error monitoring task ${taskId}:`, error);
      setTimeout(checkStatus, checkInterval);
    }
  };

  // 等待 VM 啟動後開始監控
  setTimeout(checkStatus, 10000); // 10 秒後開始檢查
}

/**
 * 更新任務狀態
 * @param {string} taskId - 任務 ID
 * @param {string} status - 新狀態
 * @param {Object} updates - 其他更新欄位
 */
function updateTaskStatus(taskId, status, updates = {}) {
  const task = tasks.get(taskId);
  if (task) {
    Object.assign(task, {
      status,
      updatedAt: Date.now(),
      ...updates,
    });
    tasks.set(taskId, task);
  }
}

/**
 * 取得任務資訊
 * @param {string} taskId - 任務 ID
 * @returns {Object|null}
 */
export function getTask(taskId) {
  return tasks.get(taskId) || null;
}

/**
 * 取得用戶的所有任務
 * @param {string} userId - LINE 用戶 ID
 * @returns {Array}
 */
export function getUserTasks(userId) {
  return Array.from(tasks.values())
    .filter(task => task.userId === userId)
    .sort((a, b) => b.createdAt - a.createdAt);
}

/**
 * 取得任務結果
 * @param {string} taskId - 任務 ID
 * @returns {Promise<Buffer|null>}
 */
export async function getTaskResult(taskId) {
  const task = tasks.get(taskId);
  if (!task || !task.resultPath) {
    return null;
  }

  // 從 GCS 路徑中提取檔案路徑
  const match = task.resultPath.match(/gs:\/\/[^/]+\/(.+)/);
  if (!match) {
    return null;
  }

  const remotePath = match[1];
  return await downloadFile(remotePath);
}

File: apps/localhost_all/src/services/task_manager.py
from typing import Dict, Optional, List
from services.vm_manager import (
    generate_instance_name, create_preemptible_vm, get_vm_status, delete_vm
)
from services.storage import upload_buffer, file_exists, download_file
from config import config
import uuid
import time

# Task status
class TaskStatus:
    PENDING = "PENDING"
    VM_CREATING = "VM_CREATING"
    VM_RUNNING = "VM_RUNNING"
    ANALYZING = "ANALYZING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    INTERRUPTED = "INTERRUPTED"


# In-memory task state (production should use Redis or database)
tasks: Dict[str, Dict] = {}


async def create_task(user_id: str, sgf_buffer: bytes, file_name: str) -> Dict:
    """Create new analysis task"""
    task_id = str(uuid.uuid4())
    timestamp = int(time.time() * 1000)
    
    # Upload SGF file to GCS
    sgf_path = f"sgf/{task_id}/{file_name}"
    sgf_gcs_path = await upload_buffer(sgf_buffer, sgf_path)
    
    # Create task record
    task = {
        "taskId": task_id,
        "userId": user_id,
        "fileName": file_name,
        "sgfPath": sgf_gcs_path,
        "status": TaskStatus.PENDING,
        "createdAt": timestamp,
        "updatedAt": timestamp,
        "instanceName": None,
        "resultPath": None,
        "error": None,
    }
    
    tasks[task_id] = task
    
    # Asynchronously start VM and execute analysis
    import asyncio
    asyncio.create_task(start_analysis(task_id))
    
    return task


async def start_analysis(task_id: str):
    """Start analysis process"""
    task = tasks.get(task_id)
    if not task:
        raise ValueError(f"Task {task_id} not found")
    
    try:
        # Generate VM instance name
        instance_name = generate_instance_name()
        update_task_status(task_id, TaskStatus.VM_CREATING, {"instanceName": instance_name})
        
        # Create Preemptible VM
        await create_preemptible_vm(instance_name, task["sgfPath"])
        update_task_status(task_id, TaskStatus.VM_RUNNING)
        
        # Start monitoring task status
        asyncio.create_task(monitor_task(task_id, instance_name))
    except Exception as error:
        logger.error(f"Failed to start analysis for task {task_id}: {error}", exc_info=True)
        update_task_status(task_id, TaskStatus.FAILED, {"error": str(error)})
        raise


async def monitor_task(task_id: str, instance_name: str):
    """Monitor task status"""
    max_wait_time = 10 * 60 * 1000  # 10 minutes
    check_interval = 5000  # Check every 5 seconds
    start_time = time.time() * 1000
    
    async def check_status():
        task = tasks.get(task_id)
        if not task:
            return
        
        # Check if timeout
        if (time.time() * 1000 - start_time) > max_wait_time:
            update_task_status(task_id, TaskStatus.FAILED, {"error": "Task timeout"})
            await delete_vm(instance_name)
            return
        
        # Check if result file exists
        result_path = f"results/{instance_name}/result.txt"
        status_path = f"results/{instance_name}/status.txt"
        
        try:
            result_exists = await file_exists(result_path)
            status_exists = await file_exists(status_path)
            
            if result_exists and status_exists:
                # Task completed
                update_task_status(
                    task_id,
                    TaskStatus.COMPLETED,
                    {"resultPath": f"gs://{config['storage']['bucket_name']}/{result_path}"}
                )
                
                # Clean up VM (if still running)
                import asyncio
                await asyncio.sleep(30)  # Wait 30 seconds before deleting
                await delete_vm(instance_name)
                return
            
            # Check VM status
            vm_status = await get_vm_status(instance_name)
            
            if not vm_status.get("exists") or vm_status.get("status") == "TERMINATED":
                # VM was interrupted or closed
                if result_exists:
                    # Has result, might be normal completion
                    update_task_status(
                        task_id,
                        TaskStatus.COMPLETED,
                        {"resultPath": f"gs://{config['storage']['bucket_name']}/{result_path}"}
                    )
                else:
                    # No result, might have been interrupted
                    update_task_status(
                        task_id,
                        TaskStatus.INTERRUPTED,
                        {"error": "VM was preempted or terminated"}
                    )
                    
                    # Retry (max 3 retries)
                    retry_count = task.get("retryCount", 0) + 1
                    if retry_count < 3:
                        print(f"Retrying task {task_id}, attempt {retry_count}")
                        task["retryCount"] = retry_count
                        await start_analysis(task_id)
                        return
                return
            
            if vm_status.get("status") == "RUNNING":
                update_task_status(task_id, TaskStatus.ANALYZING)
            
            # Continue monitoring
            await asyncio.sleep(check_interval / 1000)
            await check_status()
        except Exception as error:
            logger.error(f"Error monitoring task {task_id}: {error}", exc_info=True)
            await asyncio.sleep(check_interval / 1000)
            await check_status()
    
    # Wait for VM to start before beginning monitoring
    await asyncio.sleep(10)  # Wait 10 seconds before starting check
    await check_status()


def update_task_status(task_id: str, status: str, updates: Dict = None):
    """Update task status"""
    task = tasks.get(task_id)
    if task:
        task.update({
            "status": status,
            "updatedAt": int(time.time() * 1000),
            **(updates or {})
        })
        tasks[task_id] = task


def get_task(task_id: str) -> Optional[Dict]:
    """Get task information"""
    return tasks.get(task_id)


def get_user_tasks(user_id: str) -> List[Dict]:
    """Get all tasks for user"""
    return [
        task for task in tasks.values()
        if task.get("userId") == user_id
    ]


async def get_task_result(task_id: str) -> Optional[bytes]:
    """Get task result"""
    task = tasks.get(task_id)
    if not task or not task.get("resultPath"):
        return None
    
    # Extract file path from GCS path
    import re
    match = re.search(r"gs://[^/]+/(.+)", task["resultPath"])
    if not match:
        return None
    
    remote_path = match.group(1)
    return await download_file(remote_path)


# Add missing import
import asyncio

